### PR TITLE
Improve JAX for pretraining notebook.

### PR DESCRIPTION
**Pull request by f33ac**
## Description
This PR extends the notebook by replacing all placeholder "REVAMP" sections with actual content. The notebook ready for review initial review.

**Note:**
The vLLM section was intentionally not fully implemented, as running it inside the notebook would require more code and some hacky setup. Instead, it is briefly mentioned with guidance.

## Changes
- Replaced all "REVAMP" placeholders with content

## Testing
- Ran the notebook E2E on a `T4 GPU` & ` v5e-1 TPU` on Google Colab with no issues.

### CODE REVIEW

**Comment by f33ac** [comment]
@emilyfertig

**Comment by emilyfertig** [comment]
Hi @f33ac thanks for the PR -- I was going to start reviewing it but it looks like the file size is now 41 MB when it used to be 110 kb, do you know why that is and could you mitigate it?

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 1]
```diff
@@ -186,36 +208,184 @@
...
+    "\n",
+    "**Key features of JAX:**\n",
+    "\n",
+    "- **NumPy compatibility**: JAX provides `jax.numpy`, a drop-in replacement for NumPy that runs on accelerators like GPUs and TPUs\n",
+    "- **Automatic differentiation**: The `jax.grad()` function computes gradients automatically, making it easy to implement gradient-based optimization\n",
+    "- **JIT compilation**: `jax.jit()` compiles Python functions to optimized machine code for faster execution\n",
+    "- **Automatic parallelization**: JAX can automatically distribute computations across multiple devices using SPMD (Single Program, Multiple Data) parallelism\n",
+    "- **Functional programming**: JAX encourages pure functions, which enables reliable transformations and better performance\n",
+    "\n",
+    "For training large language models like our miniGPT, JAX's automatic parallelization capabilities are particularly valuable. They allow us to efficiently utilize multiple TPU cores without manually managing device placement and communication."
```
Comment (emilyfertig) [inline docs/source/JAX_for_LLM_pretraining.ipynb]: Let's mention GPUs here too, again to emphasize multi-platform capabilities and not imply we're TPU-only

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 2]
```diff
@@ -89,16 +123,13 @@
...
    "metadata": {
     "colab": {
      "base_uri": "https://localhost:8080/"
     },
     "id": "LS9sQEY3n0mB",
-    "outputId": "6b9ee4b0-eed0-4bae-dd99-ffed14289ad7",
-    "tags": [
-     "nbval-ignore-output"
-    ]
+    "outputId": "29c2e1dc-e49c-484a-96db-3b6c0df901a5"
```
[the remark this answers is not in the document]
Comment (emilyfertig) [reply docs/source/JAX_for_LLM_pretraining.ipynb]: Also, could you clear the outputs of this cell?

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 3]
```diff
@@ -739,814 +1046,1040 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "WDCVkgMTLC5M"
+   },
+   "source": [
+    "## Debugging the model\n",
+    "\n",
+    "When working with JAX and JIT-compiled functions, debugging requires different techniques than traditional Python debugging. Since JIT compilation traces your function and compiles it to XLA, standard `print()` statements won't work inside compiled functions.\n",
```
Comment (emilyfertig) [inline docs/source/JAX_for_LLM_pretraining.ipynb]: Add "standard `print()` statements execute at trace time and won't work inside compiled functions"

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 4]
```diff
@@ -739,814 +1046,1040 @@
...
-    "plt.plot(metrics_history['train_loss'])\n",
-    "plt.title('Training Loss')\n",
-    "plt.xlabel('Step')\n",
-    "plt.ylabel('Loss')\n",
-    "plt.show()"
+    "# Demonstrate jax.debug.print inside a JIT-compiled function\n",
+    "@jax.jit\n",
+    "def debug_forward_pass(model, inputs):\n",
+    "    \"\"\"Forward pass with debug prints to inspect intermediate values.\"\"\"\n",
+    "    # Inspect input\n",
```
Comment (emilyfertig) [inline docs/source/JAX_for_LLM_pretraining.ipynb]: Shapes are static and known at compile time, so it's reasonable to just use Python `print` for these. I think we could remove these two lines and keep the other ones that print dynamic values. (Or you could change this to a Python `print` to show the difference.)

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 5]
```diff
@@ -739,814 +1046,1040 @@
...
-    "## Debugging the model\n",
+    "# Enable NaN debugging globally - any NaN will raise an error with a traceback\n",
+    "# This is useful during development to catch numerical issues early\n",
+    "jax.config.update(\"jax_debug_nans\", True)\n",
+    "\n",
+    "# Example: This function would raise an error if it produced NaN\n",
+    "@jax.jit\n",
+    "def safe_forward_pass(model, inputs):\n",
+    "    \"\"\"Forward pass that will error if NaN is produced.\"\"\"\n",
+    "    logits = model(inputs)\n",
```
Comment (emilyfertig) [inline docs/source/JAX_for_LLM_pretraining.ipynb]: Let's add a line that actually shows the NaN check in action. After this line, you could add `logits = jnp.log(-jnp.abs(logits))`, with a comment explaining that it's a contrived example to demonstrate the NaN checker

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 6]
```diff
@@ -1601,143 +2190,54 @@
...
-      "\u001b[?25hInstalling collected packages: libclang, flatbuffers, wheel, werkzeug, tensorboardX, tensorboard-data-server, google_pasta, tensorboard, astunparse, tensorflow, qwix\n",
-      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
-      "google-tunix 0.1.3 requires datasets, which is not installed.\n",
-      "google-tunix 0.1.3 requires gcsfs, which is not installed.\n",
-      "google-tunix 0.1.3 requires hf_transfer, which is not installed.\n",
-      "google-tunix 0.1.3 requires omegaconf, which is not installed.\n",
-      "google-tunix 0.1.3 requires python-dotenv, which is not installed.\u001b[0m\u001b[31m\n",
-      "\u001b[0mSuccessfully installed astunparse-1.6.3 flatbuffers-25.9.23 google_pasta-0.2.0 libclang-18.1.1 qwix-0.1.4 tensorboard-2.20.0 tensorboard-data-server-0.7.2 tensorboardX-2.6.4 tensorflow-2.20.0 werkzeug-3.1.3 wheel-0.45.1\n"
+      "\u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/162.6 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r",
```
Comment (emilyfertig) [inline docs/source/JAX_for_LLM_pretraining.ipynb]: Clear the cell outputs?

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 7]
```diff
@@ -1748,49 +2248,69 @@
...
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 27,
    "metadata": {
-    "id": "Mtzb0NXb8TVY",
-    "tags": [
-     "nbval-ignore-output"
```
Comment (emilyfertig) [inline docs/source/JAX_for_LLM_pretraining.ipynb]: Reinstate `nbval-ignore-output`?

File: docs/source/JAX_for_LLM_pretraining.ipynb [hunk 8]
```diff
@@ -2157,162 +2720,212 @@
...
+    "- **Leveraged parallelism** with JAX's automatic SPMD for data and tensor parallelism\n",
+    "- **Debugged your model** using `jax.debug.print()` and `jax.debug_nans`\n",
+    "- **Saved checkpoints** using Orbax for model persistence\n",
+    "- **Fine-tuned with LoRA** using Tunix for parameter-efficient adaptation\n",
+    "- **Profiled performance** with Xprof to optimize hyperparameters\n",
+    "- **Explored production inference** with vLLM for high-throughput serving\n",
+    "\n",
+    "**JAX AI Stack libraries used:**\n",
+    "\n",
+    "| Library | Purpose |\n",
```
Comment (emilyfertig) [inline docs/source/JAX_for_LLM_pretraining.ipynb]: Is it possible to left-align the table? I think that'd look better in the rendering
